lib/async_hooks.js

```diff
@@ -461,6 +461,21 @@ function init(asyncId, type, triggerId, resource) {
   processing_hook = false;
 }
 
+const JSSideProviders = {
+  Timeout: 'timers Timeout',
+  Immediate: 'timers Immediate',
+  TickObject: 'next_tick TickObject'
+};
+
+function _getTypeName(module, type) {
+  return Object.keys(JSSideProviders)
+    .find((k) => JSSideProviders[k] === `${module} ${type}`);
+}
```

Comment:
> @refack
> 
> I have a better idea: lookup by function:
> As @addaleax said, if you want this to work generically only a path of pain and suffering awaits you. I spent a couple weeks trying to get this and a similar feature to work. Will never work reliably.
> 
> TBH I'm not loving adding the module name to each type. Feels like serious overkill to me.

I didn't have any bright ideas. I think that I'd invert the dependency:
```
function registerTypeName(type) {
}
```
This way implementers (or node modules) could inform us of the names they choose. We could warn about possible coalitions, but that is obviously timing dependant.